cmd/restic/cmd_recover.go

```diff
@@ -59,90 +74,113 @@ func runRecover(gopts GlobalOptions) error {
 	trees := make(map[restic.ID]bool)
 
 	for blob := range repo.Index().Each(gopts.ctx) {
-		if blob.Blob.Type != restic.TreeBlob {
-			continue
+		if blob.Type == restic.TreeBlob {
+			trees[blob.Blob.ID] = false
 		}
-		trees[blob.Blob.ID] = false
 	}
 
-	cur := 0
-	max := len(trees)
-	Verbosef("load %d trees\n\n", len(trees))
-
+	Verbosef("load %d trees\n", len(trees))
+	bar := newProgressMax(!gopts.Quiet, uint64(len(trees)), "trees loaded")
 	for id := range trees {
-		cur++
-		Verbosef("\rtree (%v/%v)", cur, max)
-
-		if !trees[id] {
-			trees[id] = false
-		}
-
 		tree, err := repo.LoadTree(gopts.ctx, id)
 		if err != nil {
 			Warnf("unable to load tree %v: %v\n", id.Str(), err)
 			continue
 		}
 
 		for _, node := range tree.Nodes {
-			if node.Type != "dir" || node.Subtree == nil {
-				continue
+			if node.Type == "dir" && node.Subtree != nil {
+				trees[*node.Subtree] = true
 			}
+		}
+		bar.Add(1)
+	}
+	bar.Done()
 
-			subtree := *node.Subtree
-			trees[subtree] = true
+	if !opts.All {
+		Verbosef("load snapshots\n")
+		snapshots, err := restic.LoadAllSnapshots(gopts.ctx, repo, nil)
+		if err != nil {
+			return err
+		}
+		for _, sn := range snapshots {
+			trees[*sn.Tree] = true
 		}
+		Verbosef("done\n")
 	}
-	Verbosef("\ndone\n")
 
 	roots := restic.NewIDSet()
 	for id, seen := range trees {
-		if seen {
-			continue
+		if !seen {
+			Verbosef("found root tree %v\n", id.Str())
```

Comment:
`s/Verbosef/Verboseff`. Printing all root tree ids is too much for normal verbosity.

Reply:
changed.